numpy/core/src/multiarray/compiled_base.dispatch.c

```diff
@@ -0,0 +1,99 @@
+/**
+ * @targets $maxopt baseline
+ * SSE2 AVX2
+ * NEON ASIMDDP
+ */
+#include "compiled_base.h"
+
+/*
+ * This function packs boolean values in the input array into the bits of a
+ * byte array. Truth values are determined as usual: 0 is false, everything
+ * else is true.
+ */
+NPY_NO_EXPORT void NPY_CPU_DISPATCH_CURFX(compiled_base_pack_inner)
+(const char *inptr, npy_intp element_size,  npy_intp n_in, npy_intp in_stride, char *outptr, npy_intp n_out, npy_intp out_stride, char order)
+{
+    /*
+     * Loop through the elements of inptr.
+     * Determine whether or not it is nonzero.
+     *  Yes: set corresponding bit (and adjust build value)
+     *  No:  move on
+     * Every 8th value, set the value of build and increment the outptr
+     */
+    npy_intp index = 0;
+    int remain = n_in % 8;              /* uneven bits */
+
+#if defined(NPY_HAVE_SSE) || defined(NPY_HAVE_AVX2) || defined(NPY_HAVE_NEON)
+    if (in_stride == 1 && element_size == 1 && n_out > 2) {
+        npyv_u64 zero = npyv_zero_u64();
+        /* don't handle non-full 8-byte remainder */
+        npy_intp vn_out = n_out - (remain ? 1 : 0);
+        const int vstep = npyv_nlanes_u64;
+        vn_out -= (vn_out & (vstep - 1));
+        for (index = 0; index < vn_out; index += vstep) {
+            // Maximum paraller abillity: handle four 64bits at one time
+            npy_uint64 a[4];
+            for (int i = 0; i < vstep; i++) {
+                a[i] = *(npy_uint64*)(inptr + 8 * i);
+                if (order == 'b') {
+                    a[i] = npy_bswap8(a[i]);
+                }
+            }
+            /* note x86 can load unaligned */
+            npyv_u64 v;
+            if (vstep == 4) {
+                v = npyv_set_u64(a[0], a[1], a[2], a[3]);
+            } else {
+                v = npyv_set_u64(a[0], a[1]);
+            }
+            /* false -> 0x00 and true -> 0xFF (there is no cmpneq) */
+            v = npyv_reinterpret_u8_u64(npyv_cmpeq_u8(npyv_reinterpret_u8_u64(v), npyv_reinterpret_u8_u64(zero)));
+            v = npyv_reinterpret_u8_u64(npyv_cmpeq_u8(npyv_reinterpret_u8_u64(v), npyv_reinterpret_u8_u64(zero)));
+            /* extract msb of 16 bytes and pack it into 16 bit */
+            unsigned int r = npyv_movemask_u8(npyv_reinterpret_u8_u64(v));
+            /* store result */
```

Comment:
```suggestion
        /**
         * TODO: load multiple vectors for each iteration and pack boolean vectors instead of movemask
         * and memcpy that will lead to a massive increase the performance 
         */
```

Reply:
a reminder mainly to me

Reply:
It's not that easy to simulate movemask in AVX512/VSX, so we'd better replace it.

Reply:
No please stay on your track packing bit will little-bit tricky(shift&bitwise) and it will require 8 or 4 loads of vectors for each iteration also you will have to add new intrinsics to handle the non-contiguous store. And after all, you will still have to implement another loop to handle small chunks of memory via single sign-mask each iteration.
I considered this comment out of topic due to the miss-confusing it can make.

> It's not that easy to simulate movemask in AVX512/VSX, so we'd better replace it.

No not at all, I did for you :)

```C
NPY_FINLINE npy_uint64 npyv_movemask_u8(npyv_b8 mask)
{
#ifdef NPY_HAVE_AVX512BW
    return (npy_uint64)_cvtmask64_u64(mask);
#else
    int mask_lo = _mm256_movemask_epi8(_mm512_castsi512_si256(mask));
    int mask_hi = _mm256_movemask_epi8(_mm512_extracti64x4_epi64(mask, 1));
    return (unsigned)mask_lo | ((npy_uint64)(unsigned)mask_hi << 32);
#endif
}

NPY_FINLINE npy_uint64 npyv_movemask_u8(npyv_b8 mask)
{
    const npyv_u8 bperm = {120, 112, 104, 96, 88, 80, 72, 64, 56, 48, 40, 32, 24, 16, 8, 0};
    npyv_s32 vmask = (npyv_s32)vec_vbpermq((npyv_u8)mask, bperm);
    return (npy_uint64)vec_extract(vmask, 2);
}
```

Reply:
Thanks for your timely help, I finished the AVX512 part but unable to finish the VSX part.